docs(secrets-management): update docs with vault rotation and some small adjustments (#5779)
Kong/docs.konghq.com: app/_src/gateway/kong-enterprise/secrets-management/secrets-rotation.md

---
title: Secrets Rotation
---

Secrets rotation is a process of updating secrets. Regular rotations of secrets is considered a good
practice. Here are some reasons why secrets rotation is important:

* Reducing impact of compromised secrets
* Enhancing resilience against brute-force attacks
* Complying with security regulations
* Maintaining separation of duties
* Adapting to evolving threats
* Mitigating the effects of insider threats

Kong needs to hold many types of secrets to be able to process requests, and to integrate with other
systems and clients:

* TLS certificates
* Credentials and API keys (to access databases, identity providers, and other services)

> * Cryptographic keys (digital signing and encryption)

Reviewer: Like in the code, I would propose that we use "rotate" to refer to the process of changing secret values on a schedule or in response to a disclosure even, and "renew" to refer to the mechanism inside of Kong that causes it to re-read secret values from their respective vaults.  What is described above is "rotation", while what is described below is "renewal".

It should also be mentioned that Kong caches secrets, as that is why the renewal mechanism is required in the first place.  It might even make sense to call the mechanism "refresh" as it is similar to OIDC refresh, so there is a precedent for using that word.

It may also be worth mentioning that in the context of vaults, the term `TTL` refers to the time that a secret is cached inside of Kong.  The term is not defined and then suddenly used, and it makes sense to provide a narrow definition to make the documentation easier to understand.

There are two main ways to rotate secrets:

* Rotate periodically using TTLs (e.g. check for new TLS certificate once per day)
* Rotate on failure (e.g. on database authentication failure, check if the secrets were updated, and try again)

Kong supports both methods for rotating secrets. Rotation on failure requires code to be written,
and thus it has limited support in Kong (Postgres credentials for now). There is an experimental
Kong PDK API that can be used to rotate secrets on failure: [kong.vault.try](/gateway/{{page.kong_version}}/plugin-development/pdk/kong.vault/#kongvaulttrycallback-options).

## Periodically rotating secrets using TTLs

> Kong employs a background job that runs periodically on **every minute** (thus you cannot currently

Reviewer: ```suggestion
Kong employs a background job that runs periodically **once every minute** (thus you cannot currently
```

automatically rotate more often than that). Its job is to rotate secrets that are about to expire.
The TTL can be configured  with the Vault (for all the secrets) or in secret reference (for a single secret).
By default, Kong does not rotate any secrets, so **remember to configure the TTLs** if you want to turn on
**the automatic rotation**.

The TTL based rotation works with most of the Kong supported vaults, including:

* [AWS Secrets Manager](/gateway/{{page.kong_version}}/kong-enterprise/secrets-management/backends/aws-sm/)
* [GCP Secrets Manager](/gateway/{{page.kong_version}}/kong-enterprise/secrets-management/backends/gcp-sm/)
* [HashiCorp Vault](/gateway/{{page.kong_version}}/kong-enterprise/secrets-management/backends/hashicorp-vault/)

> 

Reviewer: What follows seems to be a description of how the rotation process works.  I think this paragraph would better be located in the beginning of the chapter.

When rotating with TTLs, it is usually useful to have two versions of the same secret valid at the same time.
This means that following steps occur during secrets rotation:

1. A new secret (or secret version) is created, resulting in three valid ones.
2. All the three secrets are verified that they work.
3. The oldest secret (or secret version) is removed/revoked or made otherwise invalid, resulting in two valid ones.

> 

Reviewer: There is a low of copy and paste going on below.  Maybe the differences between the vaults could be in tabular form with only one copy of the shared verbiage?

### Configuring AWS Secrets Manager Secrets Rotation using TTLs

The default [AWS Secrets Manager](/gateway/{{page.kong_version}}/kong-enterprise/secrets-management/backends/aws-sm/)
vault TTLs can be configured through `kong.conf` or environment variables (the values are in seconds):

```bash
KONG_VAULT_AWS_TTL=300
KONG_VAULT_AWS_NEG_TTL=60
KONG_VAULT_AWS_RESURRECT_TTL=300
```

All AWS secret references will inherit these settings by default, e.g.:

```bash
{vault://aws/my-secret-name/foo}
```

You can override or set the TTLs directly with the references too: 

```bash
{vault://aws/my-secret-name/foo?ttl=600&neg_ttl=30&resurrect_ttl=600}
```

You can also create multiple vaults for different types of secrets,
and set the TTLs by secret types, e.g.:

{% navtabs %}
{% navtab Admin API %}

```bash
curl -i -X PUT http://HOSTNAME:8001/vaults/aws-certs  \
  --data name=aws \
  --data description="Storing secrets in AWS Secrets Manager" \
  --data config.region="us-east-1" \
  --data config.ttl=21600
```

Now when using certificates you can reference them with:

```bash
{vault://aws-certs/certs/web-site}
```

> And the all secrets (certificates in this case) referenced with `aws-certs` vault

Reviewer: ```suggestion
The secrets (certificates in this case) referenced with `aws-certs` vault
```

their expiry.

### Configuring GCP Secrets Manager Secrets Rotation using TTLs

The default [GCP Secrets Manager](/gateway/{{page.kong_version}}/kong-enterprise/secrets-management/backends/gcp-sm/)
vault TTLs can be configured through `kong.conf` or environment variables (the values are in seconds):

```bash
KONG_VAULT_GCP_TTL=300
KONG_VAULT_GCP_NEG_TTL=60
KONG_VAULT_GCP_RESURRECT_TTL=300
```

All GCP secret references will inherit these settings by default, e.g.:

```bash
{vault://gcp/my-secret-name/foo}
```

You can override or set the TTLs directly with the references too:

```bash
{vault://gcp/my-secret-name/foo?ttl=600&neg_ttl=30&resurrect_ttl=600}
```

You can also create multiple vaults for different types of secrets,
and set the TTLs by secret types, e.g.:

{% navtabs %}
{% navtab Admin API %}

```bash
curl -i -X PUT http://HOSTNAME:8001/vaults/gcp-certs  \
  --data name=gcp \
  --data description="Storing secrets in GCP Secrets Manager" \
  --data config.project_id="my_project_id-1" \
  --data config.ttl=21600
```

Now when using certificates you can reference them with:

```bash
{vault://gcp-certs/certs/web-site}
```

> And all the secrets (certificates in this case) referenced with `gcp-certs` vault

Reviewer: ```suggestion
The secrets (certificates in this case) referenced with `gcp-certs` vault
```

## Periodically rotating secrets using TTLs

> Kong automatically rotates secrets *once every minute*, through a background job. This allows uncoupling the secret rotation process from proxying. It has some consequences:
> 
> * The once-per-minute refresh rate means that a secret with a TTL of 30 seconds might take up to 60 seconds to refresh in the least favorable scenario.
> * Additionally, if the sum of TTL and `resurrect_ttl` is less than 60 seconds for a given secret, it won't be refreshed or resurrected correctly.

Reviewer 2: ```suggestion
{% if_version lte: 3.3.x %}
Kong automatically rotates secrets *once every minute* in the background. It rotates secrets that are about to expire. The TTL can be configured with the Vault (for all secrets) or in the secret reference (for a single secret). By default, Kong does not rotate any secrets, so remember to configure the TTLs if you want to enable automatic rotation.
{% endif_version %}
{% if_version gte: 3.4.x %}
Kong automatically rotates secrets *once every minute* in the background. This decouples the secret rotation process from proxying. It has the following consequences:

* The once-per-minute refresh rate means that a secret with a TTL of 30 seconds might take up to 60 seconds to refresh in the least favorable scenario.
* Additionally, if the sum of TTL and `resurrect_ttl` is less than 60 seconds for a given secret, it won't be refreshed or resurrected correctly.
{% endif_version %}
```
Added conditional rendering since the new text is only for 3.4.x or later.

## Periodically rotating secrets using TTLs

Kong employs a background job that runs periodically **once every minute** (thus you cannot currently

> automatically rotate more often than that). Its job is to rotate secrets that are about to expire.

Reviewer 3: Another attempt:

> Kong operates a background job that executes every minute, serving the function of rotating secrets that are nearing expiration. As such, it is not possible to automatically rotate secrets more frequently than this interval. This characteristic may create challenges, especially for TTLs that are set for less than one minute. The once-per-minute refresh rate means that a secret with a TTL of 30 seconds might take up to 60 seconds to refresh in the least favorable scenario. This delay arises from the alignment of the secret's TTL with the background job's minute-long execution cycle, potentially doubling the expected refresh time. Additionally, if the sum of the TTL and the resurrect_ttl is less than 60 seconds, it can lead to issues since the secret may not be refreshed or resurrected within the designated timeframe, impacting the continuity of operations that depend on the secret.

Reviewer 4: I have included a mix between this and what was there before on my final commit.

> The TTL can be configured  with the Vault (for all the secrets) or in secret reference (for a single secret).

Reviewer 3: ```suggestion
The TTL can be configured with the Vault (for all the secrets) or in secret reference (for a single secret).
```

{vault://aws-certs/certs/web-site}
```

> The secrets (certificates in this case) referenced with `aws-certs` vault

Reviewer 2: ```suggestion
The secrets (certificates, in this case) referenced with `aws-certs` vault
```

{vault://gcp-certs/certs/web-site}
```

> The secrets (certificates in this case) referenced with `gcp-certs` vault

Reviewer 2: ```suggestion
The secrets (certificates, in this case) referenced with `gcp-certs` vault
```

{vault://hcv-certs/certs/web-site}
```

> The secrets (certificates in this case) referenced with `hcv-certs` vault

Reviewer 2: ```suggestion
The secrets (certificates, in this case) referenced with `hcv-certs` vault
```
